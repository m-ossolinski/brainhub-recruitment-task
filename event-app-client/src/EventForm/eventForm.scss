.event-form {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;

  &__input {
    width: 500px;
    height: 50px;
    margin: 14px;

    &-error-msg {
      color: red;
    }
  }

  &__submit-btn {
    margin-top: 24px;
    width: 500px;
    height: 36px;
  }
}