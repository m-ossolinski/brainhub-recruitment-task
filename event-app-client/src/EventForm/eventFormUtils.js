export const EMAIL_PATTERN = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

export const eventFormConfiguration = [{
	name: 'firstName',
	id: 'firstName',
	type: 'text',
}, {
	name: 'lastName',
	id: 'lastName',
	type: 'text',
}, {
	name: 'email',
	id: 'email',
	type: 'email',
}];

const EVENT_FORM_ERROR_MESSAGES = {
	firstName: 'First name is required!',
	lastName: 'Last name is required!',
	emailRequired: 'Email is required!',
	email: 'Valid email is required!',
	eventDate: 'Event date must be selected!',
}

export const validateFormValues = (values) => {
	let errors = {};
	const {
		firstName,
		lastName,
		email,
		eventDate,
	} = values;

	console.log(values);

	if (!firstName) {
		errors.firstName = EVENT_FORM_ERROR_MESSAGES.firstName;
	}

	if (!lastName) {
		errors.lastName = EVENT_FORM_ERROR_MESSAGES.lastName;
	}

	if (!email) {
		errors.email = EVENT_FORM_ERROR_MESSAGES.emailRequired;
	}

	if (email && !EMAIL_PATTERN.test(email)) {
		errors.email = EVENT_FORM_ERROR_MESSAGES.email;
	}

	if (!eventDate) {
		errors.eventDate = EVENT_FORM_ERROR_MESSAGES.eventDate;
	}

	return errors;
}