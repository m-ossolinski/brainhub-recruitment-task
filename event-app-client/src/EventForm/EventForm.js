import React, {useEffect, useState, useRef} from 'react';
import { connect } from 'react-redux';
import DatePicker from 'react-datepicker/es';
import { eventFormConfiguration, validateFormValues, initialStateValue} from './eventFormUtils';
import { sendEventData } from '../actions/eventActions';
import './eventForm.scss';

export const EventFormComponent = (props) => {
	const {
		eventFormState: {
			sendingEventData,
			sendingEventDataFailed,
		},
		dispatch,
	} = props;

	const [eventFormValues, setEventFormValues] = useState(initialStateValue);
	const [eventFormErrors, setEventFormErrors] = useState({});
	const [eventDate, setEventDate] = useState(new Date());
	const [formStatus, setFormStatus] = useState(false);
	const firstRender = useRef(true);
console.log(eventDate);
console.log(typeof eventDate);
	const handleInputChange = e => {
		setFormStatus(false);
		const { name, value } = e.target;

		setEventFormValues({
			...eventFormValues,
			[name]: value,
		})
	}

	const handleFormSubmit = (ev) => {
		if (ev) {
			ev.preventDefault();
		}

		setEventFormErrors(validateFormValues({ ...eventFormValues, eventDate }));
	}

	useEffect(() => {
		if (firstRender.current) {
			firstRender.current = false
		} else if (Object.keys(eventFormErrors).length === 0) {
			dispatch(sendEventData({ ...eventFormValues, eventDate }));
			setFormStatus(true);
			setEventFormValues(initialStateValue);
		}
	}, [eventFormErrors])

	return (
		<div className="event-form">

			{eventFormConfiguration.map(({ name, id, type, placeholder }) => (
				<div key={id} className="event-form__element">
					<input
						name={name}
						id={id}
						type={type}
						value={eventFormValues[name]}
						onChange={handleInputChange}
						className="event-form__input"
						placeholder={placeholder}
					/>
					<span className="event-form__input-error-msg">
						{eventFormErrors[name]}
					</span>
				</div>
			))}

			<div>
				<DatePicker
					selected={eventDate}
					onChange={date => setEventDate(date)}
					dateFormat="dd/MM/yyyy"
				/>
				{eventFormErrors.eventDate ? 'Event date is required' : null}
			</div>

			<button
				onClick={handleFormSubmit}
				disabled={sendingEventData}
				className="event-form__submit-btn"
			>
				{sendingEventData ? 'Sending...' : 'Send'}
			</button>

			{sendingEventDataFailed ? 'An error occurred while submitting the form!' : null}
			{formStatus ? 'The form has been sent successfully' : null}
		</div>
	)
}

const mapStateToProps = state => ({
	eventFormState: state.event,
})

export const EventForm = connect(mapStateToProps)(EventFormComponent);