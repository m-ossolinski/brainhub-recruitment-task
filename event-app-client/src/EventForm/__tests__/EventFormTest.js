import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import TestRenderer from 'react-test-renderer';
import { act } from "react-dom/test-utils";
import { EventFormComponent } from "../EventForm";
import pretty from "pretty";
import DatePicker from "react-datepicker/es";

jest.mock('react-datepicker', () => <DatePicker />)
const dispatchMock = jest.fn();

let appContainer = null;
beforeEach(() => {
	appContainer = document.createElement("div");
	document.body.appendChild(appContainer);
});

afterEach(() => {
	unmountComponentAtNode(appContainer);
	appContainer.remove();
	appContainer = null;
});

const MOCKED_PROPS = {
	eventFormState: {
		sendingEventDataFailed: false,
		sendingEventData: false,
	},
	dispatch: dispatchMock,
}

describe('EventFormComponent', () => {
	test('should render and generate component tree properly', () => {
		act(() => {
			render(<EventFormComponent {...MOCKED_PROPS} />, appContainer)
		});

		expect(pretty(appContainer.innerHTML)).toMatchSnapshot();
	});
})

// this code also render component tree and create snapshot
// but it has been commented due to DatePicker errors

// describe('EventFormComponent snapshots', () => {
// 	test('should generate component with error about sending form fail', () => {
// 		const eventForm = TestRenderer.create(
// 			<EventFormComponent {...MOCKED_PROPS} />
// 		).toTree();
// 	});
//
// });