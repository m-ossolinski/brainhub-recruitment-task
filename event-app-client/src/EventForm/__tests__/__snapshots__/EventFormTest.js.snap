// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component should render EventFormComponent 1`] = `
"<div class=\\"event-form\\">
  <div class=\\"event-form__element\\"><input name=\\"firstName\\" id=\\"firstName\\" type=\\"text\\" class=\\"event-form__input\\" placeholder=\\"First name\\" value=\\"\\"><span class=\\"event-form__input-error-msg\\"></span></div>
  <div class=\\"event-form__element\\"><input name=\\"lastName\\" id=\\"lastName\\" type=\\"text\\" class=\\"event-form__input\\" placeholder=\\"Last name\\" value=\\"\\"><span class=\\"event-form__input-error-msg\\"></span></div>
  <div class=\\"event-form__element\\"><input name=\\"email\\" id=\\"email\\" type=\\"email\\" class=\\"event-form__input\\" placeholder=\\"E-mail\\" value=\\"\\"><span class=\\"event-form__input-error-msg\\"></span></div>
  <div></div><button class=\\"event-form__submit-btn\\">Send</button>
</div>"
`;
